src/Vulkan/Silk.NET.Vulkan/Vk.cs

```diff
@@ -232,34 +250,46 @@ public unsafe bool IsDeviceExtensionPresent(PhysicalDevice device, string extens
                 // The lack of the device handle indicates we've not been previously initialised.  We now need a write lock.
                 _cachedDeviceExtensionsLock.EnterWriteLock();
                 GlobalMemory mem = null;
+                nint layerPtr = 0;
                 try
                 {
+                    if(layer != null)
+                    {
+                        layerPtr = SilkMarshal.StringToPtr(layer);
+                    }
+
                     var deviceExtPropertiesCount = 0u;
 
                     // Get number of properties
-                    EnumerateDeviceExtensionProperties(device, (byte*) 0, &deviceExtPropertiesCount, null);
+                    EnumerateDeviceExtensionProperties(device, (byte*) layerPtr, &deviceExtPropertiesCount, null);
 
                     // Initialise return structure
-                    mem = GlobalMemory.Allocate((int) deviceExtPropertiesCount * sizeof(ExtensionProperties));
-                    var props = (ExtensionProperties*) Unsafe.AsPointer(ref mem.GetPinnableReference());
+                    var propsArr = new ExtensionProperties[(int)deviceExtPropertiesCount];
 
-                    // Get properties
-                    EnumerateDeviceExtensionProperties(device, (byte*) 0, &deviceExtPropertiesCount, props);
-                    for (int j = 0; j < deviceExtPropertiesCount; j++)
-                    {
-                        // Prefix the extension name
-                        var newKey = prefix_sep + Marshal.PtrToStringAnsi((nint) props[j].ExtensionName);
-                        _cachedDeviceExtensions.Add(newKey);
-                        if (!result && string.Equals(newKey, fullKey))
+                    fixed(ExtensionProperties* props = propsArr) {
+                        // Get properties
+                        EnumerateDeviceExtensionProperties(device, (byte*) layerPtr, &deviceExtPropertiesCount, props);
+                        for (int j = 0; j < deviceExtPropertiesCount; j++)
                         {
-                            // We found the extension (no need to do another lookup as we're scanning anyway)
-                            // As such this has taken 2 lookups + initialisation scan.
-                            result = true;
+                            // Prefix the extension name
+                            var newKey = prefixSep + Marshal.PtrToStringAnsi((nint) props[j].ExtensionName);
+                            _cachedDeviceExtensions.Add(newKey);
```

Comment:
Wow, do we really not disambiguate for each device? No need to resolve in this PR as this is an existing issue, but really this should be a `Dictionary<Device, HashSet<string>>`